### DISCUSSION TITLE
URL handler

### DISCUSSION THREAD

**Discussion by watsonSBCC** [jamulussoftware/jamulus#2607] [Ideas]
URL handler - is it possible to create a url handler so that we can post a link on a website that says jamulus://<server address="">:<port> and have it open Jamulus and connect directly to the server? Or at least just open jamulus with command line arguments.

[Apple Documentation](https://developer.apple.com/documentation/xcode/allowing_apps_and_websites_to_link_to_your_content/defining_a_custom_url_scheme_for_your_app)

Pros - easy to send Mac users (and possibly Windows users) to our department website, have them click on a link to "Jamulus Server Room A" and join the ensemble rehearsal. Makes it very easy for the users who are new to all of this and are overwhelmed with info. Just one less thing to worry about.

Cons - nightmare security risk if inputs are not parsed correctly.

**Comment by jerogee** [comment]
I _really_ like his feature request. URI handling would increase Jamulus usability, and getting choirs or bands ready to play in no time (just open the shared URI and let's jam!). 

URI handling is more developed in MacOS and Linux than in Windows but can work perfectly cross-platform. 

Some useful resources for reference:
-  Security Considerations to take into account with URIs from [RFC3986](https://tools.ietf.org/html/rfc3986#page-43) and [known attacks](https://zero.lol/2019-05-22-fun-with-uri-handlers/) that can mostly be prevented by careful input validation (argument length and permissible character sets, and force encoding)
-  Registering an [Application URL scheme in Windows](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa767914(v=vs.85)?redirectedfrom=MSDN)
- This neat article goes into further details: [Hyperlinking Beyond the Web](https://css-tricks.com/hyperlinking-beyond-the-web/)

**Comment by ann0see** [comment]
Something I would like to add for the security topic. I wouldn’t automatically connect if someone clicks on a link but maybe first show a warning which asks the user if he wants to connect to {IP}.

**Comment by ann0see** [comment]
I found this post: https://burnttoys.blogspot.com/2008/07/adding-url-scheme-to-qt-application.html?m=1

It might help

**Comment by djfun** [comment]
The URL handling itself shouldn't be too difficult to implement. The question is: what should happen if Jamulus is already running? Can we get the URL to the already running instance? This would be especially useful, if the user starts the client with extra command line arguments.

**Comment by ann0see** [comment]
Hi, Sorry for the maintenance noise here. I’m just into triaging issues.


Does anyone want to take this issue up? I think it should be quite doable in terms of implementation but needs some research.

This issue needs some more discussion and is not actionable yet. I’ve therefore moved it to a GitHub discussion to maintain more visibility. Please note that GitHub Discussions – different to GitHub issues show up high if there is discussion. 

Note: I will unsubscribe from this issue and won’t receive responses from any new comments. If you have any questions concerning maintenance (re-opening this issue if you can’t), feel free to ping me.
